calico_node/startup/startup.go

```diff
@@ -312,35 +341,75 @@ func validateIP(ip *net.IP) {
 
 // autoDetectCIDR auto-detects the IP and Network using the requested
 // detection method.
-func autoDetectCIDR(detectionMethod string, version int) *net.IP {
+func autoDetectCIDR(method string, version int) *net.IP {
+	if method == "" || method == AUTOD_METHOD_FIRST {
+		// Autodetect the IP by enumerating all interfaces (excluding
+		// known internal interfaces).
+		return autoDetectCIDRFirstFound(version)
+	} else if strings.HasPrefix(method, AUTOD_METHOD_INTERFACE) {
+		// Autodetect the IP from the specified interface.
+		ifStr := strings.TrimPrefix(method, AUTOD_METHOD_INTERFACE)
+		return autoDetectCIDRByInterface(ifStr, version)
+	} else if strings.HasPrefix(method, AUTOD_METHOD_CAN_REACH) {
+		// Autodetect the IP by connecting a UDP socket to a supplied address.
+		destStr := strings.TrimPrefix(method, AUTOD_METHOD_CAN_REACH)
+		return autoDetectCIDRByReach(destStr, version)
+	}
+
+	// The autodetection method is not recognised and is required.  Exit.
+	fatal("Invalid IP autodection method: %s", method)
+	terminate()
+	return nil
+}
+
+// autoDetectCIDRFirstFound auto-detects the first valid Network it finds across
+// all interfaces (excluding common known internal interface names).
+func autoDetectCIDRFirstFound(version int) *net.IP {
+
 	incl := []string{}
-	excl := []string{"^docker.*", "^cbr.*", "dummy.*",
+	excl := []string{
+		"docker.*", "cbr.*", "dummy.*",
```

Comment:
curious why you removed `^` from the regex. Wondering someone with some modification to docker0 bridge can have an interface like `...-docker...` and if this will change the behavior we use to have.